From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rezoss <rezoss79@users.noreply.github.com>
Date: Sat, 11 Jul 2026 00:00:00 +0200
Subject: [PATCH] Enable Now Nudge chat and autofill compatibility

Keep chat fields in the Now Nudge flow, extend WhatsApp and Telegram policy
coverage, recover WhatsApp compose nodes, and tolerate missing user context.

---
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/interpreter/AssistDataInterpreter.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/interpreter/AssistDataInterpreter.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/interpreter/AssistDataInterpreter.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/interpreter/AssistDataInterpreter.smali
@@ -642,6 +642,230 @@
     return p0
 .end method
 
+.method private final shouldSkipUnsupportedChatFieldHide(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;)Z
+    .locals 2
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;->getNodeInfo()Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getAutofillType()I
+
+    move-result p0
+
+    const/4 v0, 0x1
+
+    if-ne p0, v0, :cond_not_chat_package
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;->getComponentName()Landroid/content/ComponentName;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Landroid/content/ComponentName;->getPackageName()Ljava/lang/String;
+
+    move-result-object p0
+
+    const-string p1, "com.whatsapp"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.whatsapp.w4b"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.google.android.apps.messaging"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.google.android.apps.dynamite"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.kakao.talk"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "jp.naver.line.android"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "org.thoughtcrime.securesms"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.instagram.android"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.sgiggle.production"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.nttdocomo.android.msg"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.kddi.android.cmail"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.verizon.messaging.vzmsgs"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "org.telegram.messenger"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "org.telegram.messenger.web"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.samsung.android.messaging"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.facebook.orca"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.tencent.mm"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.viber.voip"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.discord"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.snapchat.android"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.microsoft.teams"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "com.google.android.apps.googlevoice"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "org.telegram.plus"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_chat_package
+
+    const-string p1, "org.thunderdog.challegram"
+
+    invoke-static {p0, p1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result p0
+
+    if-nez p0, :cond_chat_package
+
+    :cond_not_chat_package
+    const/4 p0, 0x0
+
+    return p0
+
+    :cond_chat_package
+    return v0
+.end method
+
 .method private final isFocusedFieldSupported(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;)Z
     .locals 1
 
@@ -1294,6 +1518,22 @@
 
     if-nez v0, :cond_0
 
+    invoke-direct {p0, p1}, Lcom/samsung/android/smartsuggestions/service/autofill/interpreter/AssistDataInterpreter;->shouldSkipUnsupportedChatFieldHide(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_hide_unsupported_field
+
+    const-string v0, "Unsupported chat field; continue Now Nudge flow"
+
+    invoke-static {v1, v0}, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillFlowLogger;->i(Ljava/lang/String;Ljava/lang/String;)V
+
+    const/4 v0, 0x1
+
+    goto :cond_0
+
+    :cond_hide_unsupported_field
+
     const-string p1, "Focused field is not supported"
 
     invoke-static {v1, p1}, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillFlowLogger;->i(Ljava/lang/String;Ljava/lang/String;)V

diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/AutofillNudgeManager.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/AutofillNudgeManager.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/AutofillNudgeManager.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/AutofillNudgeManager.smali
@@ -515,11 +515,9 @@
 
     invoke-static {v4, p1}, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillFlowLogger;->e(Ljava/lang/String;Ljava/lang/String;)V
 
-    sget-object p1, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillNudgeHelper;->INSTANCE:Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillNudgeHelper;
-
-    iget-object p0, p0, Lcom/samsung/android/smartsuggestions/service/autofill/AutofillNudgeManager;->context:Landroid/content/Context;
+    const-string p0, "Missing autofill node; keep Now Nudge flow alive"
 
-    invoke-virtual {p1, p0}, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillNudgeHelper;->hideInlineSuggestion(Landroid/content/Context;)V
+    invoke-static {v4, p0}, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillFlowLogger;->i(Ljava/lang/String;Ljava/lang/String;)V
 
     :cond_3
     return-void

diff --git a/smali_classes14/com/samsung/android/smartsuggestions/extservice/services/ContentCaptureService.smali b/smali_classes14/com/samsung/android/smartsuggestions/extservice/services/ContentCaptureService.smali
--- a/smali_classes14/com/samsung/android/smartsuggestions/extservice/services/ContentCaptureService.smali
+++ b/smali_classes14/com/samsung/android/smartsuggestions/extservice/services/ContentCaptureService.smali
@@ -403,6 +403,167 @@
 
     move-result-object v0
 
+    if-eqz v0, :cond_rezoss_chat_capture_allowlist_done
+
+    const-string v1, "com.whatsapp"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.whatsapp.w4b"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "org.telegram.messenger"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "org.telegram.messenger.web"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "org.telegram.plus"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "org.thunderdog.challegram"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.facebook.orca"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.google.android.apps.messaging"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.snapchat.android"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.samsung.android.messaging"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.zing.zalo"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.viber.voip"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.imo.android.imoim"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.discord"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "ru.oneme.app"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.kakao.talk"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.tencent.mm"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "org.thoughtcrime.securesms"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.imo.android.imoimhd"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.beint.zangi"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.imo.android.imoimbeta"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "org.aka.messenger"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.imo.android.imoimlite"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.michatapp.im"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.loudtalks"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.facebook.talk"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "ir.ilmili.telegraph"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.vk.im"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.turkcell.bip"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.groupme.android"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.azarlive.android"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.microsoft.teams"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.ss.android.ugc.trill"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.facebook.katana"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.instagram.android"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.tencent.mobileqq"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.samsung.vvm"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.sds.teams"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.google.android.apps.dynamite"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "SSS:ContentCaptureService"
+
+    :cond_rezoss_chat_capture_allowlist_done
     invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/policy/Policylist;->getContentCaptureAllowlist()Lcom/samsung/android/smartsuggestions/service/policy/Allowlist;
 
     move-result-object p1
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/policy/Policylist.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/policy/Policylist.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/policy/Policylist.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/policy/Policylist.smali
@@ -160,10 +160,43 @@
 
 # virtual methods
 .method public getAppViewIdEntries()Ljava/lang/String;
-    .locals 0
+    .locals 2
 
     iget-object p0, p0, Lcom/samsung/android/smartsuggestions/service/policy/Policylist;->mAppViewIdEntries:Ljava/lang/String;
 
+    if-eqz p0, :cond_rezoss_return_app_view_id_entries
+
+    const-string v0, "org.telegram.messenger"
+
+    invoke-virtual {p0, v0}, Ljava/lang/String;->contains(Ljava/lang/CharSequence;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_rezoss_return_app_view_id_entries
+
+    invoke-virtual {p0}, Ljava/lang/String;->length()I
+
+    move-result v0
+
+    const/4 v1, 0x1
+
+    if-le v0, v1, :cond_rezoss_return_app_view_id_entries
+
+    add-int/lit8 v0, v0, -0x1
+
+    const/4 v1, 0x0
+
+    invoke-virtual {p0, v1, v0}, Ljava/lang/String;->substring(II)Ljava/lang/String;
+
+    move-result-object p0
+
+    const-string/jumbo v0, ";\"org.telegram.messenger\":{\"message_IdEntry\":\"message\";\"message_IdEntry_extension\":[\"message_text\";\"text\";\"text1\";\"text2\";\"instant_message\";\"shortcut_widget_item_message\"];\"title_IdEntry\":\"action_bar_title\";\"title_IdEntry_extension\":[\"title\";\"title_view\";\"chat_title\";\"profile_name\";\"name\"];\"listview_IdEntry\":\"list_view\";\"listview_IdEntry_extension\":[\"messages_list\";\"chat_list_view\";\"chat_list\"]};\"org.telegram.messenger.web\":{\"message_IdEntry\":\"message\";\"message_IdEntry_extension\":[\"message_text\";\"text\";\"text1\";\"text2\";\"instant_message\";\"shortcut_widget_item_message\"];\"title_IdEntry\":\"action_bar_title\";\"title_IdEntry_extension\":[\"title\";\"title_view\";\"chat_title\";\"profile_name\";\"name\"];\"listview_IdEntry\":\"list_view\";\"listview_IdEntry_extension\":[\"messages_list\";\"chat_list_view\";\"chat_list\"]};\"org.telegram.plus\":{\"message_IdEntry\":\"message\";\"message_IdEntry_extension\":[\"message_text\";\"text\";\"text1\";\"text2\";\"instant_message\";\"shortcut_widget_item_message\"];\"title_IdEntry\":\"action_bar_title\";\"title_IdEntry_extension\":[\"title\";\"title_view\";\"chat_title\";\"profile_name\";\"name\"];\"listview_IdEntry\":\"list_view\";\"listview_IdEntry_extension\":[\"messages_list\";\"chat_list_view\";\"chat_list\"]};\"org.thunderdog.challegram\":{\"message_IdEntry\":\"message\";\"message_IdEntry_extension\":[\"message_text\";\"text\";\"text1\";\"text2\";\"instant_message\";\"shortcut_widget_item_message\"];\"title_IdEntry\":\"action_bar_title\";\"title_IdEntry_extension\":[\"title\";\"title_view\";\"chat_title\";\"profile_name\";\"name\"];\"listview_IdEntry\":\"list_view\";\"listview_IdEntry_extension\":[\"messages_list\";\"chat_list_view\";\"chat_list\"]}}"
+
+    invoke-virtual {p0, v0}, Ljava/lang/String;->concat(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object p0
+
+    :cond_rezoss_return_app_view_id_entries
     return-object p0
 .end method
 
@@ -417,6 +450,162 @@
     goto :goto_1
 
     :cond_0
+    const-string p0, "com.whatsapp"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.whatsapp.w4b"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "org.telegram.messenger"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "org.telegram.messenger.web"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "org.telegram.plus"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "org.thunderdog.challegram"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.facebook.orca"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.google.android.apps.messaging"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.snapchat.android"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.samsung.android.messaging"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.zing.zalo"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.viber.voip"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.imo.android.imoim"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.discord"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "ru.oneme.app"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.kakao.talk"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.tencent.mm"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "org.thoughtcrime.securesms"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.imo.android.imoimhd"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.beint.zangi"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.imo.android.imoimbeta"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "org.aka.messenger"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.imo.android.imoimlite"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.michatapp.im"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.loudtalks"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.facebook.talk"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "ir.ilmili.telegraph"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.vk.im"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.turkcell.bip"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.groupme.android"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.azarlive.android"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.microsoft.teams"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.ss.android.ugc.trill"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.facebook.katana"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.instagram.android"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.tencent.mobileqq"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.samsung.vvm"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.sds.teams"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "com.google.android.apps.dynamite"
+
+    invoke-virtual {v0, p0}, Ljava/util/ArrayList;->add(Ljava/lang/Object;)Z
+
     return-object v0
 
     :goto_1

diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder.smali
@@ -82,6 +82,136 @@
     return-void
 .end method
 
+.method private final findWhatsappComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Landroid/app/assist/AssistStructure$ViewNode;
+    .locals 4
+
+    if-eqz p1, :cond_rezoss_not_found
+
+    invoke-direct {p0, p1}, Lcom/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder;->isWhatsappComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_rezoss_check_children
+
+    const-string v0, "AssistStructureFinder"
+
+    const-string v1, "Recovered WhatsApp compose node for collectAssistData"
+
+    invoke-static {v0, v1}, Lxc/e;->c(Ljava/lang/String;Ljava/lang/String;)V
+
+    invoke-direct {p0, p1}, Lcom/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder;->printViewNode(Landroid/app/assist/AssistStructure$ViewNode;)V
+
+    return-object p1
+
+    :cond_rezoss_check_children
+    invoke-virtual {p1}, Landroid/app/assist/AssistStructure$ViewNode;->getChildCount()I
+
+    move-result v0
+
+    const/4 v1, 0x0
+
+    :goto_rezoss_child_loop
+    if-ge v1, v0, :cond_rezoss_not_found
+
+    invoke-virtual {p1, v1}, Landroid/app/assist/AssistStructure$ViewNode;->getChildAt(I)Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object v2
+
+    invoke-direct {p0, v2}, Lcom/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder;->findWhatsappComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object v3
+
+    if-eqz v3, :cond_rezoss_next_child
+
+    return-object v3
+
+    :cond_rezoss_next_child
+    add-int/lit8 v1, v1, 0x1
+
+    goto :goto_rezoss_child_loop
+
+    :cond_rezoss_not_found
+    const/4 p0, 0x0
+
+    return-object p0
+.end method
+
+.method private final isWhatsappComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Z
+    .locals 3
+
+    invoke-virtual {p1}, Landroid/app/assist/AssistStructure$ViewNode;->getIdEntry()Ljava/lang/String;
+
+    move-result-object v0
+
+    const-string v1, "entry"
+
+    invoke-static {v0, v1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_rezoss_true
+
+    if-eqz v0, :cond_rezoss_check_text_id
+
+    invoke-static {v0, v1}, Lkotlin/text/v;->l0(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_rezoss_check_class
+
+    :cond_rezoss_check_text_id
+    invoke-virtual {p1}, Landroid/app/assist/AssistStructure$ViewNode;->getTextIdEntry()Ljava/lang/String;
+
+    move-result-object v0
+
+    invoke-static {v0, v1}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_rezoss_true
+
+    if-eqz v0, :cond_rezoss_check_class
+
+    invoke-static {v0, v1}, Lkotlin/text/v;->l0(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_rezoss_check_class
+
+    :cond_rezoss_check_class
+    invoke-virtual {p1}, Landroid/app/assist/AssistStructure$ViewNode;->getAutofillType()I
+
+    move-result p0
+
+    const/4 v2, 0x1
+
+    if-ne p0, v2, :cond_rezoss_false
+
+    invoke-virtual {p1}, Landroid/app/assist/AssistStructure$ViewNode;->getClassName()Ljava/lang/String;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_rezoss_false
+
+    const-string v1, "EditText"
+
+    invoke-static {v0, v1}, Lkotlin/text/v;->l0(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_rezoss_true
+
+    :cond_rezoss_false
+    const/4 p0, 0x0
+
+    return p0
+
+    :cond_rezoss_true
+    const/4 p0, 0x1
+
+    return p0
+.end method
+
 .method private final findNode(Ljava/lang/String;Landroid/app/assist/AssistStructure$ViewNode;)Landroid/app/assist/AssistStructure$ViewNode;
     .locals 4
 
@@ -233,7 +337,7 @@
 
 # virtual methods
 .method public final find(Landroid/content/ComponentName;Ljava/lang/String;)Landroid/app/assist/AssistStructure$ViewNode;
-    .locals 5
+    .locals 7
     .param p1    # Landroid/content/ComponentName;
         .annotation build Lorg/jetbrains/annotations/NotNull;
         .end annotation
@@ -253,6 +357,30 @@
 
     invoke-static {p2, v0}, Lkotlin/jvm/internal/h;->q(Ljava/lang/Object;Ljava/lang/String;)V
 
+    invoke-virtual {p1}, Landroid/content/ComponentName;->getPackageName()Ljava/lang/String;
+
+    move-result-object v5
+
+    const-string v0, "com.whatsapp"
+
+    invoke-static {v5, v0}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_rezoss_whatsapp_package
+
+    const-string v0, "com.whatsapp.w4b"
+
+    invoke-static {v5, v0}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v5
+
+    goto :goto_rezoss_package_checked
+
+    :cond_rezoss_whatsapp_package
+    const/4 v5, 0x1
+
+    :goto_rezoss_package_checked
     invoke-static {}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/assist/AssistHolder;->getInstance()Lcom/samsung/android/smartsuggestions/service/deepscreencapture/assist/AssistHolder;
 
     move-result-object v0
@@ -310,6 +436,8 @@
 
     invoke-static {v2}, Lkotlin/jvm/internal/h;->n(Ljava/lang/Object;)V
 
+    move-object v6, v2
+
     invoke-direct {p0, p2, v2}, Lcom/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder;->findNode(Ljava/lang/String;Landroid/app/assist/AssistStructure$ViewNode;)Landroid/app/assist/AssistStructure$ViewNode;
 
     move-result-object v2
@@ -324,6 +452,21 @@
     return-object v2
 
     :cond_1
+    if-eqz v5, :cond_rezoss_next_window
+
+    invoke-direct {p0, v6}, Lcom/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder;->findWhatsappComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object v2
+
+    if-eqz v2, :cond_rezoss_next_window
+
+    const-string p0, "!!! whatsapp compose fallback find"
+
+    invoke-static {v4, p0}, Lxc/e;->c(Ljava/lang/String;Ljava/lang/String;)V
+
+    return-object v2
+
+    :cond_rezoss_next_window
     add-int/lit8 v1, v1, 0x1
 
     goto :goto_1

diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController.smali
@@ -781,6 +781,94 @@
 
     return p0
 .end method
+
+.method private final isWhatsappEntry(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;)Z
+    .locals 4
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;->getComponentName()Landroid/content/ComponentName;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Landroid/content/ComponentName;->getPackageName()Ljava/lang/String;
+
+    move-result-object p0
+
+    const-string v0, "com.whatsapp"
+
+    invoke-static {p0, v0}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_rezoss_package_ok
+
+    const-string v0, "com.whatsapp.w4b"
+
+    invoke-static {p0, v0}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_rezoss_false
+
+    :cond_rezoss_package_ok
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;->getNodeInfo()Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object p0
+
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getAutofillType()I
+
+    move-result p1
+
+    const/4 v2, 0x1
+
+    if-ne p1, v2, :cond_rezoss_false
+
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getIdEntry()Ljava/lang/String;
+
+    move-result-object p1
+
+    const-string v3, "entry"
+
+    invoke-static {p1, v3}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_rezoss_true
+
+    if-eqz p1, :cond_rezoss_check_text_id
+
+    invoke-static {p1, v3}, Lkotlin/text/v;->l0(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result p1
+
+    if-nez p1, :cond_rezoss_true
+
+    :cond_rezoss_check_text_id
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getTextIdEntry()Ljava/lang/String;
+
+    move-result-object p0
+
+    invoke-static {p0, v3}, Lkotlin/jvm/internal/h;->j(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result p1
+
+    if-nez p1, :cond_rezoss_true
+
+    if-eqz p0, :cond_rezoss_false
+
+    invoke-static {p0, v3}, Lkotlin/text/v;->l0(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result p0
+
+    if-nez p0, :cond_rezoss_true
+
+    :cond_rezoss_false
+    const/4 p0, 0x0
+
+    return p0
+
+    :cond_rezoss_true
+    return v2
+.end method
 
 .method private final matched(Landroid/view/autofill/AutofillId;Ljava/util/Set;)Z
     .locals 3
@@ -921,6 +1009,33 @@
 
     :cond_0
-    sget-object p1, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillNudgeHelper;->INSTANCE:Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillNudgeHelper;
+    invoke-direct {p0, p1}, Lcom/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController;->isWhatsappEntry(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_rezoss_hide_inline
+
+    new-instance v0, Ljava/util/LinkedHashMap;
+
+    invoke-direct {v0}, Ljava/util/LinkedHashMap;-><init>()V
+
+    sget-object v1, Lcom/samsung/android/smartsuggestions/service/autofill/data/FieldInputType;->UNKNOWN:Lcom/samsung/android/smartsuggestions/service/autofill/data/FieldInputType;
+
+    const-string v2, "message"
+
+    invoke-interface {v0, v1, v2}, Ljava/util/Map;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v1, "AutofillEventController"
+
+    const-string v2, "WhatsApp entry TEXT has no PDE userContext; synthesize chat/message context"
+
+    invoke-static {v1, v2}, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillFlowLogger;->i(Ljava/lang/String;Ljava/lang/String;)V
+
+    invoke-direct {p0, p1, v0}, Lcom/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController;->showNudge(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;Ljava/util/Map;)V
+
+    return-void
+
+    :cond_rezoss_hide_inline
+    sget-object p1, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillNudgeHelper;->INSTANCE:Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillNudgeHelper;
 
     iget-object p0, p0, Lcom/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController;->context:Landroid/content/Context;
 
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/ResourceIdEntityMap.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/ResourceIdEntityMap.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/ResourceIdEntityMap.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/ResourceIdEntityMap.smali
@@ -137,5 +137,23 @@
     goto :goto_0
 
     :cond_1
+    const-string v1, "com.whatsapp.Conversation_com.whatsapp_entry"
+
+    const-string/jumbo v2, "message"
+
+    invoke-virtual {v0, v1, v2}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v1, "com.whatsapp.Conversation_com.whatsapp.w4b_entry"
+
+    invoke-virtual {v0, v1, v2}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v1, "com.whatsapp.w4b.Conversation_com.whatsapp.w4b_entry"
+
+    invoke-virtual {v0, v1, v2}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string v1, "com.whatsapp.w4b.Conversation_com.whatsapp_entry"
+
+    invoke-virtual {v0, v1, v2}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
     return-object v0
 .end method
